### PR TITLE
feat: Wasm extension HTTP code source

### PR DESCRIPTION
**Pull request by zhaohuabing**
This PR implements the HTTP code source for Wasm extension.

### CODE REVIEW

File: api/v1alpha1/wasm_types.go [hunk 1]
```diff
@@ -78,8 +78,9 @@ type WasmCodeSource struct {
 	Image *ImageWasmCodeSource `json:"image,omitempty"`
 
 	// SHA256 checksum that will be used to verify the wasm code.
-	// +optional
-	// SHA256 *string `json:"sha256,omitempty"`
+	//
+	// kubebuilder:validation:Pattern=`^[a-f0-9]{64}$`
+	SHA256 string `json:"sha256"`
```
Comment (zhaohuabing) [inline api/v1alpha1/wasm_types.go:83]: Make SHA256 mandatory for now to avoid downloading wasm and calculate SHA in the control plane.

We can make it optional in the future without breaking changes if we opt to build a cache in the control plane.

File: api/v1alpha1/envoyextensionypolicy_types.go [hunk 2]
```diff
@@ -46,12 +46,12 @@ type EnvoyExtensionPolicySpec struct {
...
 	TargetRef gwapiv1a2.PolicyTargetReferenceWithSectionName `json:"targetRef"`
 
-	// WASM is a list of Wasm extensions to be loaded by the Gateway.
+	// Wasm is a list of Wasm extensions to be loaded by the Gateway.
 	// Order matters, as the extensions will be loaded in the order they are
 	// defined in this list.
 	//
 	// +optional
-	WASM []Wasm `json:"wasm,omitempty"`
+	Wasm []Wasm `json:"wasm,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/envoyextensionypolicy_types.go:54]: Struggled between `WASM` and `Wasm`. Looks like `Wasm` is [the official abbreviation](https://webassembly.org/). :-)

Comment (Xunzhuo) [reply api/v1alpha1/envoyextensionypolicy_types.go:54]: It looks a bit confused, the name is Wasm and the type is the slice of Wasm ([]Wasm)

Comment (zhaohuabing) [reply api/v1alpha1/envoyextensionypolicy_types.go:54]: Yeah, it's not pretty. I'm following the existing ExtProc setting here.

Alternative:

```
Wasms []Wasm `json:"wasm,omitempty"`

ExtProcs []ExtProc `json:"extProc,omitempty"`
```

```
WasmList []Wasm `json:"wasm,omitempty"`

ExtProcList []ExtProc `json:"extProc,omitempty"`
```

@arkodg @guydc

Comment (zhaohuabing) [reply api/v1alpha1/envoyextensionypolicy_types.go:54]: Follow existing EG API types to use `Wasms` and `ExtProcs`.

Comment (arkodg) [reply api/v1alpha1/envoyextensionypolicy_types.go]: We are using acronyms here, let's use Wasm and ExtProc

**Review by guydc — APPROVED** [review]
LGTM 

Would be great to have a minimal e2e test for one of the sources to demonstrate functional correctness.

File: api/v1alpha1/wasm_types.go [hunk 3]
```diff
@@ -31,14 +31,16 @@ type Wasm struct {
 	// RootID is a unique ID for a set of extensions in a VM which will share a
 	// RootContext and Contexts if applicable (e.g., an Wasm HttpFilter and an Wasm AccessLog).
 	// If left blank, all extensions with a blank root_id with the same vm_id will share Context(s).
-	// RootID *string `json:"rootID,omitempty"`
+	// RootID must match the root_id parameter used to register the Context in the Wasm code.
+	RootID *string `json:"rootID,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/wasm_types.go:35]: This is needed and can't be generated by EG. See: https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/921039ae983ce053bf5cba78a85a3c08ff9791e5/proxy_wasm_intrinsics.cc#L27-L28

File: test/e2e/testdata/wasm.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: http-with-http-wasm-source
+  wasm:
+  - name: wasm-filter
+    rootID: my_root_id
+    code:
+      type: HTTP
+      http:
+        url: https://raw.githubusercontent.com/envoyproxy/envoy/main/examples/wasm-cc/lib/envoy_filter_http_wasm_example.wasm
```
Comment (zhaohuabing) [inline test/e2e/testdata/wasm.yaml:54]: @arkodg The HTTP code source does support HTTPS schema.

File: api/v1alpha1/wasm_types.go [hunk 5]
```diff
@@ -92,6 +95,9 @@ const (
 
 	// ImageWasmCodeSourceType allows the user to specify the wasm code in an OCI image.
 	ImageWasmCodeSourceType WasmCodeSourceType = "Image"
+
+	// ConfigMapCodeSourceType allows the user to specify the wasm code in a ConfigMap.
```
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:99]: can we move this into a separate PR ?

File: api/v1alpha1/wasm_types.go (hunk 3 above)
Comment (arkodg) [inline api/v1alpha1/wasm_types.go:35]: why can't we reuse `name` internally here ?

Comment (arkodg) [reply api/v1alpha1/wasm_types.go:35]: If we can generate this field internally, using name, my vote is to do that and not expose another field the user needs to think about how to populate

Comment (zhaohuabing) [reply api/v1alpha1/wasm_types.go:35]: We can't generate `RootID`, See: https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/921039ae983ce053bf5cba78a85a3c08ff9791e5/proxy_wasm_intrinsics.cc#L27-L28

File: api/v1alpha1/envoyextensionypolicy_types.go [hunk 6]
```diff
@@ -46,12 +46,12 @@ type EnvoyExtensionPolicySpec struct {
 	// TargetRef
 	TargetRef gwapiv1a2.PolicyTargetReferenceWithSectionName `json:"targetRef"`
 
-	// WASM is a list of Wasm extensions to be loaded by the Gateway.
+	// Wasm is a list of Wasm extensions to be loaded by the Gateway.
```
[the remark this answers is not in the document]
Comment (zhaohuabing) [reply api/v1alpha1/envoyextensionypolicy_types.go:49]: Following the name convention here：https://webassembly.org/

```
WebAssembly (abbreviated Wasm) is a binary instruction format for a stack-based virtual machine. Wasm is designed as a portable compilation target for programming languages, enabling deployment on the web for client and server applications.
```

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zirain — APPROVED** [review]
